types/index.d.ts

```diff
@@ -0,0 +1,265 @@
+  type ReplyFnResult =
+    | readonly [StatusCode]
+    | readonly [StatusCode, ReplyBody]
+    | readonly [StatusCode, ReplyBody, ReplyHeaders]
+
+  interface ReplyFnContext extends Interceptor {
+    req: ClientRequest & {
+      headers: Record<string, string>
+    }
+  }
+
+  interface Scope extends NodeJS.EventEmitter {
+    get: InterceptFunction
+    post: InterceptFunction
+    put: InterceptFunction
+    head: InterceptFunction
+    patch: InterceptFunction
+    merge: InterceptFunction
+    delete: InterceptFunction
+    options: InterceptFunction
+
+    intercept: (
+      uri: string | RegExp | { (uri: string): boolean },
+      method: string,
+      requestBody?: RequestBodyMatcher,
+      options?: Options
+    ) => Interceptor
+
+    defaultReplyHeaders(headers: ReplyHeaders): this
+    matchHeader(name: string, value: RequestHeaderMatcher): this
+    filteringPath(regex: RegExp, replace: string): this
+    filteringPath(fn: (path: string) => string): this
+    filteringRequestBody(regex: RegExp, replace: string): this
+    filteringRequestBody(fn: (body: string) => string): this
+
+    log(out: (message: any, ...optionalParams: any[]) => void): this
+    persist(flag?: boolean): this
+    replyContentLength(): this
+    replyDate(d?: Date): this
+
+    done(): void
+    isDone(): boolean
+    restore(): void
+    pendingMocks(): string[]
+    activeMocks(): string[]
+  }
+
+  interface Interceptor {
+    query(
+      matcher:
+        | boolean
+        | string
+        | DataMatcher
+        | URLSearchParams
+        | { (parsedObj: ParsedUrlQuery): boolean }
+    ): this
+
+    // tslint (as of 5.16) is under the impression that the callback types can be unified,
+    // however, doing so causes the params to lose their inherited types during use.
+    // the order of the overrides is important for determining the param types in the replay fns.
+    /* tslint:disable:unified-signatures */
+    reply(
+      replyFnWithCallback: (
+        this: ReplyFnContext,
+        uri: string,
+        body: Body,
+        callback: (
+          err: NodeJS.ErrnoException | null,
+          result: ReplyFnResult
+        ) => void
+      ) => void
+    ): Scope
+    reply(
+      replyFn: (
+        this: ReplyFnContext,
+        uri: string,
+        body: Body
+      ) => ReplyFnResult | Promise<ReplyFnResult>
+    ): Scope
+    reply(
+      statusCode: StatusCode,
+      replyBodyFnWithCallback: (
+        this: ReplyFnContext,
+        uri: string,
+        body: Body,
+        callback: (err: NodeJS.ErrnoException | null, result: ReplyBody) => void
+      ) => void,
+      headers?: ReplyHeaders
+    ): Scope
+    reply(
+      statusCode: StatusCode,
+      replyBodyFn: (
+        this: ReplyFnContext,
+        uri: string,
+        body: Body
+      ) => ReplyBody | Promise<ReplyBody>,
+      headers?: ReplyHeaders
+    ): Scope
+    reply(responseCode?: StatusCode, body?: Body, headers?: ReplyHeaders): Scope
```

Comment:
Do you know if we accept e.g. `.reply(undefined, "foobar")` as this indicates?

Reply:
I had to double check, but yes. That would work and result in the status code defaulting to 200.